dev/io.openliberty.microprofile.telemetry.1.0.internal_fat/fat/src/io/openliberty/microprofile/telemetry/internal_fat/apps/jaxrspropagation/JaxRsEndpoints.java

```diff
@@ -331,6 +341,75 @@ public Response getJax(@Context UriInfo uriInfo) {
         return Response.ok(Span.current().getSpanContext().getTraceId()).build();
     }
 
+    @GET
+    @Path("/contextPropagatorTwo")
+    public Response contextPropagatorTwo(@Context HttpHeaders headers) throws Exception {
+        String traceState = Span.current().getSpanContext().getTraceState().toString();
+        if(traceState.contains("entries=[test, tracestate]")){
+            return Response.ok("Tracestate found").build();
+        }
+        return Response.ok("Tracestate not found").build();
+    }
+
+    @GET
+    @Path("/traceState")
+    public Response testTraceState(@Context UriInfo uriInfo) throws Exception {
+        try (Scope s = Baggage.builder().put("foo", "bar").build().makeCurrent()) {
+            Span parentSpan = Span.current();
+            TraceState parentState = parentSpan.getSpanContext().getTraceState();
+            assertNotNull(parentState);
+
+            String url = new String(uriInfo.getAbsolutePath().toString());
+            url = url.replace("traceState", "contextPropagatorTwo"); //The jaxrsclient will use the URL as given so it needs the final part to be provided.
+            Client client = ClientBuilder.newClient();
+            WebTarget resourceTarget = client.target(url);
+            Invocation.Builder builder = resourceTarget.request();
+
+            TextMapPropagator textMapPropagator = openTelemetry.getPropagators().getTextMapPropagator();
+            openTelemetry.getPropagators().getTextMapPropagator().inject(withSpanContext(
+                                                         SpanContext.create(
+                                                                            parentSpan.getSpanContext().getTraceId(),
+                                                                            parentSpan.getSpanContext().getSpanId(),
+                                                                            TraceFlags.getSampled(),
+                                                                            TraceState.builder().put("test", "tracestate").build()),
+                                                         io.opentelemetry.context.Context.current()),
+                                                                         builder, new JaxRsRequestSetter());
+
+            assertEquals("Tracestate found", builder.get(String.class));
+            return Response.ok(Span.current().getSpanContext().getTraceId()).build();
```

Comment:
There are a couple of ways you could get a `Span` with a `TraceState` here.

1. You could manually create the `tracestate` header in `JaxRsIntegration.testTraceState`. Liberty/Otel should turn it into a `TraceState` object on the current span by the time it gets to this method.
    This is probably the best and simplest option.
2. You could manually create a span with a tracestate with `Span.wrap(SpanContext.create(traceid, spanid, traceFlags, traceState))` and set it as the current span. That span won't get traced, but it can be used as the parent of another span (such as the one created when you make the REST request) and the new span should have the same tracestate.

Reply:
The W3C Trace Context recommendation [describes the format of the tracestate header](https://www.w3.org/TR/trace-context/#tracestate-header)